File: backend/app/services/estimates.py
from __future__ import annotations

from app.schemas.estimate import EstimateCreate, EstimateLineItem, EstimateLineItemCost, EstimateSummary


def calculate_estimate(payload: EstimateCreate) -> EstimateSummary:
    line_item_costs = [calculate_line_item(item) for item in payload.line_items]
    direct_cost = sum(item.direct_cost for item in line_item_costs)
    indirect_cost = sum(indirect.amount for indirect in payload.indirects)
    risk_cost = sum(risk.amount for risk in payload.risks)
    subtotal_before_markup = direct_cost + indirect_cost + risk_cost
    contingency = subtotal_before_markup * payload.markup.contingency_percent / 100
    overhead_base = subtotal_before_markup + contingency
    overhead = overhead_base * payload.markup.overhead_percent / 100
    profit_base = overhead_base + overhead
    profit = profit_base * payload.markup.profit_percent / 100
    final_price = profit_base + profit
    gross_margin_percent = ((final_price - direct_cost) / final_price * 100) if final_price else 0

    return EstimateSummary(
        project_name=payload.project_name,
        project_code=payload.project_code,
        direct_cost=round(direct_cost, 2),
        indirect_cost=round(indirect_cost, 2),
        risk_cost=round(risk_cost, 2),
        subtotal_before_markup=round(subtotal_before_markup, 2),
        contingency=round(contingency, 2),
        overhead=round(overhead, 2),
        profit=round(profit, 2),
        final_price=round(final_price, 2),
        gross_margin_percent=round(gross_margin_percent, 2),
        line_items=line_item_costs,
        assumptions=payload.assumptions,
        exclusions=payload.exclusions,
    )


def calculate_line_item(item: EstimateLineItem) -> EstimateLineItemCost:
    hours = _calculate_hours(item)
    labour_cost = sum(
        member.quantity * member.burdened_hourly_rate * hours for member in item.labour
    )
    equipment_cost = sum(resource.quantity * resource.hourly_rate * hours for resource in item.equipment)
    material_cost = sum(
        material.quantity * material.unit_cost * (1 + material.waste_percent / 100)
        for material in item.materials
    )
    subcontract_cost = item.subcontract.quoted_amount if item.subcontract else 0
    explicit_direct_cost = item.direct_unit_cost * item.quantity if item.direct_unit_cost is not None else 0
    direct_cost = labour_cost + equipment_cost + material_cost + subcontract_cost + explicit_direct_cost
    unit_cost = direct_cost / item.quantity if item.quantity else 0

    return EstimateLineItemCost(
        description=item.description,
        item_type=item.item_type,
        quantity=item.quantity,
        unit=item.unit,
        hours=round(hours, 2),
        labour_cost=round(labour_cost, 2),
        equipment_cost=round(equipment_cost, 2),
        material_cost=round(material_cost, 2),
        subcontract_cost=round(subcontract_cost, 2),
        direct_cost=round(direct_cost, 2),
        unit_cost=round(unit_cost, 2),
    )


def _calculate_hours(item: EstimateLineItem) -> float:
    if item.production_rate_per_hour:
        return item.quantity / item.production_rate_per_hour
    return 0

File: frontend/src/api/estimates.ts
export type EstimateUnit = "LS" | "EA" | "m" | "m2" | "m3" | "t" | "hr" | "day";

export type EstimateItemType = "self_perform" | "material" | "subcontract" | "indirect" | "allowance";

export type DefaultProductionActivity =
  | "pipe_installation"
  | "excavation"
  | "bedding"
  | "backfill"
  | "asphalt_removal"
  | "concrete_removal"
  | "manhole_installation"
  | "catch_basin_installation"
  | "sidewalk"
  | "curb"
  | "traffic_control"
  | "landscaping";

export type LabourCrewMember = {
  role: string;
  quantity: number;
  hourly_rate: number;
  burden_percent: number;
};

export type EquipmentResource = {
  name: string;
  quantity: number;
  hourly_rate: number;
  daily_rate?: number | null;
  owned_or_rented?: string | null;
};

export type MaterialInput = {
  name: string;
  quantity: number;
  unit: EstimateUnit;
  unit_cost: number;
  supplier?: string | null;
  waste_percent: number;
};

export type VendorQuoteInput = {
  supplier: string;
  scope: string;
  amount: number;
  is_selected: boolean;
  notes?: string | null;
};

export type EstimateLineItem = {
  code?: string | null;
  description: string;
  item_type: EstimateItemType;
  quantity: number;
  unit: EstimateUnit;
  production_rate_per_hour?: number | null;
  default_activity?: DefaultProductionActivity | null;
  labour: LabourCrewMember[];
  equipment: EquipmentResource[];
  materials: MaterialInput[];
  vendor_quotes: VendorQuoteInput[];
  direct_unit_cost?: number | null;
  notes?: string | null;
};

export type EstimateIndirect = {
  description: string;
  amount: number;
  category?: string | null;
};

export type EstimateRiskAllowance = {
  description: string;
  amount: number;
  probability?: number | null;
  notes?: string | null;
};

export type EstimateMarkup = {
  overhead_percent: number;
  profit_percent: number;
  contingency_percent: number;
  bonding_percent: number;
  insurance_percent: number;
};

export type EstimateCreate = {
  project_name: string;
  project_code?: string | null;
  owner?: string | null;
  estimator?: string | null;
  line_items: EstimateLineItem[];
  indirects: EstimateIndirect[];
  risks: EstimateRiskAllowance[];
  markup: EstimateMarkup;
  assumptions: string[];
  exclusions: string[];
};

export type TakeoffHandoffItem = {
  code?: string | null;
  description: string;
  category: string;
  quantity: number;
  unit: EstimateUnit;
  source?: string | null;
  confidence: number;
  drawing_reference?: string | null;
  notes?: string | null;
};

export type EstimateHandoffRequest = {
  project_name: string;
  project_code?: string | null;
  items: TakeoffHandoffItem[];
};

export type EstimateHandoffResponse = {
  project_name: string;
  project_code?: string | null;
  line_items: EstimateLineItem[];
  warnings: string[];
  assumptions: string[];
};

export type EstimateLineItemCost = {
  code?: string | null;
  description: string;
  item_type: EstimateItemType;
  quantity: number;
  unit: EstimateUnit;
  hours: number;
  labour_cost: number;
  equipment_cost: number;
  material_cost: number;
  disposal_cost: number;
  subcontract_cost: number;
  direct_cost: number;
  unit_cost: number;
  selected_quote_supplier?: string | null;
  selected_quote_amount?: string | null;
};

export type EstimateCategoryBreakdown = {
  labour: number;
  equipment: number;
  material: number;
  disposal: number;
  subcontract: number;
  indirect: number;
  risk: number;
};

export type EstimateSummary = {
  project_name: string;
  project_code?: string | null;
  direct_cost: number;
  indirect_cost: number;
  risk_cost: number;
  subtotal_before_markup: number;
  contingency: number;
  bonding: number;
  insurance: number;
  overhead: number;
  profit: number;
  final_price: number;
  gross_margin_percent: number;
  category_breakdown: EstimateCategoryBreakdown;
  line_items: EstimateLineItemCost[];
  assumptions: string[];
  exclusions: string[];
};

export type ProductionRate = {
  activity: DefaultProductionActivity;
  description: string;
  unit: EstimateUnit;
  production_rate_per_hour: number;
  crew: LabourCrewMember[];
  equipment: EquipmentResource[];
  notes?: string | null;
};

export type RateLibrary = {
  production_rates: ProductionRate[];
};

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL ?? "http://localhost:8000/api/v1";

async function request<T>(path: string, options?: RequestInit): Promise<T> {
  const response = await fetch(`${API_BASE_URL}${path}`, {
    headers: {
      "Content-Type": "application/json",
      ...options?.headers,
    },
    ...options,
  });

  if (!response.ok) {
    throw new Error(`Request failed with ${response.status}`);
  }

  return response.json() as Promise<T>;
}

export const estimatesApi = {
  rateLibrary: () => request<RateLibrary>("/estimates/rate-library"),
  handoff: (payload: EstimateHandoffRequest) =>
    request<EstimateHandoffResponse>("/estimates/handoff", {
      method: "POST",
      body: JSON.stringify(payload),
    }),
  summary: (payload: EstimateCreate) =>
    request<EstimateSummary>("/estimates/summary", {
      method: "POST",
      body: JSON.stringify(payload),
    }),
  workbookUrl: () => `${API_BASE_URL}/estimates/workbook`,
};